File: katas/php/OOP/abstraction/AbstractionTests.php
<?php

use PHPUnit\Framework\TestCase;

require_once 'abstraction-good.php';

/**
 * Tests para el concepto de Abstracción en OOP
 * Valida que la abstracción oculte detalles de implementación
 */
class AbstractionTests extends TestCase
{
    public function testEmailSender_CanBeInstantiated()
    {
        // Arrange & Act
        $sender = new EmailSender("usuario@gmail.com", "password123");
        
        // Assert
        $this->assertNotNull($sender, 'El EmailSender debe poder ser instanciado');
    }
    
    public function testEmailSender_SendsEmailSuccessfully()
    {
        // Arrange
        $sender = new EmailSender("usuario@gmail.com", "password123");
        
        // Act
        $result = $sender->sendEmail("destinatario@gmail.com", "Hola", "Mensaje de prueba");
        
        // Assert
        $this->assertTrue($result, 'El email debe enviarse exitosamente');
    }
    
    public function testEmailSender_FailsWithEmptyCredentials()
    {
        // Arrange
        $sender = new EmailSender("", "");
        
        // Act
        $result = $sender->sendEmail("destinatario@gmail.com", "Hola", "Mensaje");
        
        // Assert
        $this->assertFalse($result, 'Debe fallar al enviar email con credenciales vacías');
    }
    
    public function testEmailSender_AbstractsComplexity()
    {
        // Arrange
        $sender = new EmailSender("usuario@gmail.com", "password123");
        
        // Act - Un solo método simple abstrae toda la complejidad
        $result = $sender->sendEmail("test@test.com", "Test", "Test body");
        
        // Assert
        $this->assertTrue($result, 'La abstracción debe simplificar el proceso de envío de email');
    }
}

File: katas/java/OOP/abstraction/AbstractionTests.java
package OOP.abstraction;

import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

/**
 * Tests para el concepto de Abstracción en OOP
 * Valida que la abstracción oculte detalles de implementación
 */
public class AbstractionTests {
    
    @Test
    public void testEmailSender_CanBeInstantiated() {
        // Arrange & Act
        EmailSender sender = new EmailSender("usuario@gmail.com", "password123");
        
        // Assert
        assertNotNull(sender, "El EmailSender debe poder ser instanciado");
    }
    
    @Test
    public void testEmailSender_SendsEmailSuccessfully() {
        // Arrange
        EmailSender sender = new EmailSender("usuario@gmail.com", "password123");
        
        // Act
        boolean result = sender.sendEmail("destinatario@gmail.com", "Hola", "Mensaje de prueba");
        
        // Assert
        assertTrue(result, "El email debe enviarse exitosamente");
    }
    
    @Test
    public void testEmailSender_FailsWithEmptyCredentials() {
        // Arrange
        EmailSender sender = new EmailSender("", "");
        
        // Act
        boolean result = sender.sendEmail("destinatario@gmail.com", "Hola", "Mensaje");
        
        // Assert
        assertFalse(result, "Debe fallar al enviar email con credenciales vacías");
    }
    
    @Test
    public void testEmailSender_AbstractsComplexity() {
        // Arrange
        EmailSender sender = new EmailSender("usuario@gmail.com", "password123");
        
        // Act - Un solo método simple abstrae toda la complejidad
        boolean result = sender.sendEmail("test@test.com", "Test", "Test body");
        
        // Assert
        assertTrue(result, "La abstracción debe simplificar el proceso de envío de email");
    }
}
